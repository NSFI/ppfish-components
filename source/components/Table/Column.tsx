import * as React from 'react';
import { ColumnProps } from './interface';

export default class Column<T> extends React.Component<ColumnProps<T>, React.ComponentState> {}
