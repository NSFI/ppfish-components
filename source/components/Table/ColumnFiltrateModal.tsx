import React from 'react';
import { polyfill } from 'react-lifecycles-compat';

import Modal from '../Modal';
import Icon from '../Icon';
import Checkbox from '../Checkbox';
import ToolTip from '../Tooltip';
import { ColumnFiltrateProps, ColumnFiltrateState } from './interface';
import ConfigConsumer from '../Config/Consumer';
const CheckboxGroup = Checkbox.Group;

const getColumnKey = column => {
  if ('colSpan' in column) {
    return column.filtrateTitle || column.title;
  } else {
    return column.key || column.dataIndex || column.filtrateTitle || column.title;
  }
};

const flatten = arr => {
  return arr.reduce(function (prev, next) {
    return prev.concat(Array.isArray(next) ? flatten(next) : next);
  }, []);
};

const getCheckedOptions = (columns, hideColumns) => {
  return (
    columns
      // 去除表头合并不显示的列
      .filter(column => column.colSpan !== 0)
      // 去除初始化就被隐藏的列
      .filter(column => hideColumns.findIndex(key => key === getColumnKey(column)) === -1)
      // 只留uniqueId
      .map(column => getColumnKey(column))
  );
};

const getColSpanOption = columns => {
  const colSpanOption = [];
  columns.forEach((column, index) => {
    if ('colSpan' in column && column.colSpan !== 0) {
      colSpanOption.push({
        key: column.filtrateTitle || column.title,
        value: columns
          .slice(index, index + column.colSpan)
          .map(column => column.key || column.dataIndex),
      });
    }
  });
  return colSpanOption;
};

class ColumnFiltrateModal<T> extends React.Component<ColumnFiltrateProps<T>, ColumnFiltrateState> {
  static defaultProps = {
    defaultColumns: [],
    hideColumns: [],
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    const { prevProps = {} } = prevState;
    const newState: any = { prevProps: nextProps };
    //监听column个数变化，重新初始化checkedOption
    if ('columns' in nextProps && nextProps.columns.length !== prevProps.columns.length) {
      const { columns, hideColumns } = nextProps;
      const option = getCheckedOptions(columns, hideColumns);
      newState.checkedOption = option;
      newState.checkedOptionConfirm = option;
      newState.colSpanOption = getColSpanOption(columns);
    }
    return newState;
  }

  constructor(props: ColumnFiltrateProps<T>) {
    super(props);
    const { columns, hideColumns } = this.props;
    const option = getCheckedOptions(columns, hideColumns);
    this.state = {
      visible: false,
      checkedOption: option,
      checkedOptionConfirm: option,
      okButtonDisabled: true,
      prevProps: props,
      colSpanOption: getColSpanOption(columns),
    };
  }

  showModal = () => {
    this.setState({
      visible: true,
    });
  };

  handleOk = () => {
    this.setState(
      {
        visible: false,
        checkedOptionConfirm: this.state.checkedOption,
        okButtonDisabled: true,
      },
      () => {
        // 被隐藏的列表key
        this.props.onChange(
          flatten(
            this.getOptions()
              .filter(column => this.state.checkedOptionConfirm.indexOf(column.value) === -1)
              .map(column => {
                const indexOfColSpan = this.state.colSpanOption.findIndex(
                  option => option.key === column.value,
                );
                if (indexOfColSpan !== -1) {
                  return this.state.colSpanOption[indexOfColSpan].value;
                } else {
                  return column.value;
                }
              }),
          ),
        );
      },
    );
  };

  handleCancel = () => {
    this.setState({
      visible: false,
      checkedOption: this.state.checkedOptionConfirm,
      okButtonDisabled: true,
    });
  };

  onChange = checkedOption => {
    let { checkedOptionConfirm } = this.state;
    this.setState({
      checkedOption,
      okButtonDisabled:
        JSON.stringify(checkedOption && checkedOption.sort()) ==
        JSON.stringify(checkedOptionConfirm && checkedOptionConfirm.sort()),
    });
  };

  getOptions = () => {
    const { columns, defaultColumns } = this.props;
    return (
      columns
        // 去除表头合并的不显示的
        .filter(column => column.colSpan !== 0)
        // 获取他们的label、value、disabled
        .map(column => {
          const uniqKey = getColumnKey(column);
          //fixed /分组不能控制显示隐藏
          const disabled = !!column.fixed || !!column.children || !!column.filtrateDefault;
          const title = column.filtrateTitle || column.title;
          return {
            label: <ToolTip title={title}>{title}</ToolTip>,
            value: uniqKey,
            disabled: disabled,
          };
        })
        // 去除默认展示数据项
        .filter(column => defaultColumns.findIndex(key => key === column.value) === -1)
    );
  };

  //默认一直显示的选项
  getDefaultOption = () => {
    const { columns, defaultColumns = [] } = this.props;
    return columns
      .filter(column => defaultColumns.findIndex(key => key === getColumnKey(column)) !== -1)
      .map(column => ({
        label: (
          <ToolTip title={column.filtrateTitle || column.title}>
            {column.filtrateTitle || column.title}
          </ToolTip>
        ),
        value: getColumnKey(column),
        disabled: true,
      }));
  };

  render() {
    const { prefixCls } = this.props;
    const { visible, checkedOption, okButtonDisabled } = this.state;
    const options = this.getOptions();
    const defaultOption = this.getDefaultOption();
    return (
      <ConfigConsumer componentName="Table">
        {Locale => {
          return (
            <div>
              <Icon
                className={`${prefixCls}-filtrate-icon`}
                type="Settingx"
                onClick={this.showModal}
              />
              <Modal
                title={Locale.modalTitle}
                wrapClassName={`${prefixCls}-filtrate-modal`}
                visible={visible}
                width={680}
                okButtonDisabled={okButtonDisabled}
                onOk={this.handleOk}
                onCancel={this.handleCancel}
                okText={Locale.okText}
                cancelText={Locale.cancelText}
              >
                <div className={`${prefixCls}-filtrate-modal-enable-list`}>
                  <CheckboxGroup value={checkedOption} options={options} onChange={this.onChange} />
                </div>
                {defaultOption && !!defaultOption.length && (
                  <div className={`${prefixCls}-filtrate-modal-disabled-list`}>
                    <p className="title">{Locale.defaultOptionTitle}</p>
                    <CheckboxGroup
                      options={defaultOption}
                      defaultValue={defaultOption.map(option => option.value)}
                    />
                  </div>
                )}
              </Modal>
            </div>
          );
        }}
      </ConfigConsumer>
    );
  }
}

polyfill(ColumnFiltrateModal);

export default ColumnFiltrateModal;
