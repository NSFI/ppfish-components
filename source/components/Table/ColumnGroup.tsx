import * as React from 'react';

export interface ColumnGroupProps {
  title?: React.ReactNode;
}

export default class ColumnGroup extends React.Component<ColumnGroupProps, React.ComponentState> {
  static __ANT_TABLE_COLUMN_GROUP = true;
}
