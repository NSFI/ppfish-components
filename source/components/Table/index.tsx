import Table from './Table';

export * from './interface';

export default Table;
