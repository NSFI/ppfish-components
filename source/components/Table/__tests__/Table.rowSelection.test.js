import React from 'react';
import { mount, render } from 'enzyme';
import Table from '../index.tsx';
import Checkbox from '../../Checkbox/index.tsx';

describe('Table.rowSelection', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  const columns = [
    {
      title: 'Name',
      dataIndex: 'name',
    },
  ];

  const data = [
    { key: 0, name: 'Jack' },
    { key: 1, name: 'Lucy' },
    { key: 2, name: 'Tom' },
    { key: 3, name: 'Jerry' },
  ];

  function createTable(props = {}) {
    return <Table columns={columns} dataSource={data} rowSelection={{}} {...props} />;
  }

  function renderedNames(wrapper) {
    return wrapper.find('TableRow').map(row => row.props().record.name);
  }

  it('select by checkbox', () => {
    const wrapper = mount(createTable());
    const checkboxes = wrapper.find('input');
    const checkboxAll = checkboxes.first();

    checkboxAll.simulate('change', { target: { checked: true } });
    expect(wrapper.instance().store.getState()).toEqual({
      selectedRowKeys: [0, 1, 2, 3],
      selectionDirty: true,
    });

    checkboxes.at(1).simulate('change', { target: { checked: false } });
    expect(wrapper.instance().store.getState()).toEqual({
      selectedRowKeys: [1, 2, 3],
      selectionDirty: true,
    });

    checkboxes.at(1).simulate('change', { target: { checked: true } });
    expect(wrapper.instance().store.getState()).toEqual({
      selectedRowKeys: [1, 2, 3, 0],
      selectionDirty: true,
    });
  });

  it('select by radio', () => {
    const wrapper = mount(createTable({ rowSelection: { type: 'radio' } }));
    const radios = wrapper.find('input');

    expect(radios.length).toBe(4);

    radios.first().simulate('change', { target: { checked: true } });
    expect(wrapper.instance().store.getState()).toEqual({
      selectedRowKeys: [0],
      selectionDirty: true,
    });

    radios.last().simulate('change', { target: { checked: true } });
    expect(wrapper.instance().store.getState()).toEqual({
      selectedRowKeys: [3],
      selectionDirty: true,
    });
  });

  it('pass getCheckboxProps to checkbox', () => {
    const rowSelection = {
      getCheckboxProps: record => ({
        disabled: record.name === 'Lucy',
        name: record.name,
      }),
    };

    const wrapper = mount(createTable({ rowSelection }));
    const checkboxes = wrapper.find('input');

    expect(checkboxes.at(1).props().disabled).toBe(false);
    expect(checkboxes.at(1).props().name).toEqual(data[0].name);
    expect(checkboxes.at(2).props().disabled).toBe(true);
    expect(checkboxes.at(2).props().name).toEqual(data[1].name);
  });

  it('works with pagination', () => {
    const wrapper = mount(createTable({ pagination: { pageSize: 2 } }));

    const checkboxAll = wrapper.find('SelectionCheckboxAll');
    const pagers = wrapper.find('Pager');

    checkboxAll.find('input').simulate('change', { target: { checked: true } });
    expect(checkboxAll.instance().state).toEqual({ checked: true, indeterminate: false });

    pagers.at(1).simulate('click');
    expect(checkboxAll.instance().state).toEqual({ checked: false, indeterminate: false });

    pagers.at(0).simulate('click');
    expect(checkboxAll.instance().state).toEqual({ checked: true, indeterminate: false });
  });

  // https://github.com/ant-design/ant-design/issues/4020
  it('handles defaultChecked', () => {
    const rowSelection = {
      getCheckboxProps: record => ({
        defaultChecked: record.key === 0,
      }),
    };

    const wrapper = mount(createTable({ rowSelection }));

    let checkboxs = wrapper.find('input');
    expect(checkboxs.at(1).props().checked).toBe(true);
    expect(checkboxs.at(2).props().checked).toBe(false);

    checkboxs.at(2).simulate('change', { target: { checked: true } });
    checkboxs = wrapper.find('input');
    expect(checkboxs.at(1).props().checked).toBe(true);
    expect(checkboxs.at(2).props().checked).toBe(true);
  });

  it('can be controlled', () => {
    const wrapper = mount(createTable({ rowSelection: { selectedRowKeys: [0] } }));

    expect(wrapper.instance().store.getState()).toEqual({
      selectedRowKeys: [0],
      selectionDirty: false,
    });

    wrapper.setProps({ rowSelection: { selectedRowKeys: [1] } });

    expect(wrapper.instance().store.getState()).toEqual({
      selectedRowKeys: [1],
      selectionDirty: false,
    });
  });

  it('fires change & select events', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const rowSelection = {
      onChange: handleChange,
      onSelect: handleSelect,
    };
    const wrapper = mount(createTable({ rowSelection }));

    wrapper
      .find('input')
      .last()
      .simulate('change', { target: { checked: true } });

    expect(handleChange).toBeCalledWith([3], [{ key: 3, name: 'Jerry' }]);
    expect(handleSelect.mock.calls.length).toBe(1);
    expect(handleSelect.mock.calls[0][0]).toEqual({ key: 3, name: 'Jerry' });
    expect(handleSelect.mock.calls[0][1]).toEqual(true);
    expect(handleSelect.mock.calls[0][2]).toEqual([{ key: 3, name: 'Jerry' }]);
    expect(handleSelect.mock.calls[0][3].type).toBe('change');
  });

  it('fires selectAll event', () => {
    const handleSelectAll = jest.fn();
    const rowSelection = {
      onSelectAll: handleSelectAll,
    };
    const wrapper = mount(createTable({ rowSelection }));

    wrapper
      .find('input')
      .first()
      .simulate('change', { target: { checked: true } });
    expect(handleSelectAll).toBeCalledWith(true, data, data);

    wrapper
      .find('input')
      .first()
      .simulate('change', { target: { checked: false } });
    expect(handleSelectAll).toBeCalledWith(false, [], data);
  });

  it('render with default selection correctly', () => {
    const rowSelection = {
      selections: true,
    };
    const wrapper = render(createTable({ rowSelection }));
    expect(wrapper).toMatchSnapshot();
  });

  it('click select all selection', () => {
    // jest.useFakeTimers();
    const handleSelectAll = jest.fn();
    const rowSelection = {
      onSelectAll: handleSelectAll,
      selections: true,
    };
    const wrapper = mount(createTable({ rowSelection }));

    // Open
    wrapper.find('.fishd-table-selection-down').simulate('mouseEnter');

    jest.runAllTimers();
    wrapper.update();

    wrapper.find('.fishd-dropdown-menu-item div').first().simulate('click');

    expect(handleSelectAll).toBeCalledWith(true, data, data);
  });

  it('fires selectInvert event', () => {
    // jest.useFakeTimers();
    const handleSelectInvert = jest.fn();
    const rowSelection = {
      onSelectInvert: handleSelectInvert,
      selections: true,
    };
    const wrapper = mount(createTable({ rowSelection }));
    const checkboxes = wrapper.find('input');

    checkboxes.at(1).simulate('change', { target: { checked: true } });

    // Open
    wrapper.find('.fishd-table-selection-down').simulate('mouseEnter');

    jest.runAllTimers();
    wrapper.update();

    wrapper.find('.fishd-dropdown-menu-item div').last().simulate('click');

    expect(handleSelectInvert).toBeCalledWith([1, 2, 3]);
  });

  it('fires selection event', () => {
    // jest.useFakeTimers();
    const handleSelectOdd = jest.fn();
    const handleSelectEven = jest.fn();
    const rowSelection = {
      selections: [
        {
          key: 'odd',
          text: '奇数项',
          onSelect: handleSelectOdd,
        },
        {
          key: 'even',
          text: '偶数项',
          onSelect: handleSelectEven,
        },
      ],
    };
    const wrapper = mount(createTable({ rowSelection }));

    // Open
    wrapper.find('.fishd-table-selection-down').simulate('mouseEnter');

    jest.runAllTimers();
    wrapper.update();

    // const dropdownWrapper = mount(wrapper.find('Trigger').first().instance().getComponent());
    expect(wrapper.find('.fishd-dropdown-menu-item').hostNodes().length).toBe(4);

    wrapper.find('.fishd-dropdown-menu-item').hostNodes().at(2).find('div').simulate('click');
    expect(handleSelectOdd).toBeCalledWith([0, 1, 2, 3]);

    wrapper.find('.fishd-dropdown-menu-item').hostNodes().at(3).find('div').simulate('click');
    expect(handleSelectEven).toBeCalledWith([0, 1, 2, 3]);
  });

  it('could hide default selection options', () => {
    // jest.useFakeTimers();
    const rowSelection = {
      hideDefaultSelections: true,
      selections: [
        {
          key: 'odd',
          text: '奇数项',
        },
        {
          key: 'even',
          text: '偶数项',
        },
      ],
    };
    const wrapper = mount(createTable({ rowSelection }));
    // Open
    wrapper.find('.fishd-table-selection-down').simulate('mouseEnter');

    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.fishd-dropdown-menu-item').hostNodes().length).toBe(2);
  });

  it('handle custom selection onSelect correctly when hide default selection options', () => {
    // jest.useFakeTimers();
    const handleSelectOdd = jest.fn();
    const handleSelectEven = jest.fn();
    const rowSelection = {
      hideDefaultSelections: true,
      selections: [
        {
          key: 'odd',
          text: '奇数项',
          onSelect: handleSelectOdd,
        },
        {
          key: 'even',
          text: '偶数项',
          onSelect: handleSelectEven,
        },
      ],
    };
    const wrapper = mount(createTable({ rowSelection }));

    // Open
    wrapper.find('.fishd-table-selection-down').simulate('mouseEnter');

    jest.runAllTimers();
    wrapper.update();

    wrapper.find('.fishd-dropdown-menu-item div').at(0).simulate('click');
    expect(handleSelectOdd).toBeCalledWith([0, 1, 2, 3]);

    wrapper.find('.fishd-dropdown-menu-item div').at(1).simulate('click');
    expect(handleSelectEven).toBeCalledWith([0, 1, 2, 3]);
  });

  // https://github.com/ant-design/ant-design/issues/4245
  it('handles disabled checkbox correctly when dataSource changes', () => {
    const rowSelection = {
      getCheckboxProps: record => ({ disabled: record.disabled }),
    };
    const wrapper = mount(createTable({ rowSelection }));
    const newData = [
      { key: 0, name: 'Jack', disabled: true },
      { key: 1, name: 'Lucy', disabled: true },
    ];
    wrapper.setProps({ dataSource: newData });
    wrapper.find('input').forEach(checkbox => {
      expect(checkbox.props().disabled).toBe(true);
    });
  });

  // https://github.com/ant-design/ant-design/issues/4245
  it('should allow dynamic getCheckboxProps', () => {
    class App extends React.Component {
      state = {
        disableName: 'Jack',
      };

      render() {
        const { disableName } = this.state;
        return (
          <Table
            columns={columns}
            dataSource={data}
            rowSelection={{
              getCheckboxProps: record => ({ disabled: record.name === disableName }),
            }}
          />
        );
      }
    }

    const wrapper = mount(<App />);
    let checkboxs = wrapper.find('input');
    expect(checkboxs.at(1).props().disabled).toBe(true);
    expect(checkboxs.at(2).props().disabled).toBe(false);
    wrapper.setState({ disableName: 'Lucy' });
    checkboxs = wrapper.find('input');
    expect(checkboxs.at(1).props().disabled).toBe(false);
    expect(checkboxs.at(2).props().disabled).toBe(true);
  });

  // https://github.com/ant-design/ant-design/issues/4779
  it('should not switch pagination when select record', () => {
    const newData = [];
    for (let i = 0; i < 20; i += 1) {
      newData.push({
        key: i.toString(),
        name: i.toString(),
      });
    }
    const wrapper = mount(
      createTable({
        rowSelection: {},
        dataSource: newData,
      }),
    );
    wrapper.find('Pager').last().simulate('click'); // switch to second page
    wrapper
      .find('input')
      .first()
      .simulate('change', { target: { checked: true } });
    wrapper.update();
    expect(renderedNames(wrapper)).toEqual([
      '10',
      '11',
      '12',
      '13',
      '14',
      '15',
      '16',
      '17',
      '18',
      '19',
    ]);
  });

  it('highlight selected row', () => {
    const wrapper = mount(createTable());
    wrapper
      .find('input')
      .at(1)
      .simulate('change', { target: { checked: true } });
    expect(wrapper.find('tbody tr').at(0).hasClass('fishd-table-row-selected')).toBe(true);
  });

  it('fix selection column on the left', () => {
    const wrapper = render(
      createTable({
        rowSelection: { fixed: true },
      }),
    );

    expect(wrapper).toMatchSnapshot();
  });

  // https://github.com/ant-design/ant-design/issues/10629
  it('should keep all checked state when remove item from dataSource', () => {
    const wrapper = mount(
      <Table
        rowSelection={{
          selectedRowKeys: [0, 1, 2, 3],
        }}
        columns={columns}
        dataSource={data}
      />,
    );
    expect(wrapper.find(Checkbox).length).toBe(5);
    wrapper.find(Checkbox).forEach(checkbox => {
      expect(checkbox.props().checked).toBe(true);
      expect(checkbox.props().indeterminate).toBe(false);
    });
    wrapper.setProps({
      dataSource: data.slice(1),
      rowSelection: {
        selectedRowKeys: [1, 2, 3],
      },
    });
    expect(wrapper.find(Checkbox).length).toBe(4);
    wrapper.find(Checkbox).forEach(checkbox => {
      expect(checkbox.props().checked).toBe(true);
      expect(checkbox.props().indeterminate).toBe(false);
    });
  });
});
