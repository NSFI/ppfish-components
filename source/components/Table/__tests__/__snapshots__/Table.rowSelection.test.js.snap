// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.rowSelection fix selection column on the left 1`] = `
<div
  class="fishd-table-wrapper"
>
  <div
    class="fishd-spin-nested-loading"
  >
    <div
      class="fishd-spin-nested"
    >
      <div>
        <div
          class="fishd-table fishd-table-default fishd-table-scroll-position-left"
        >
          <div
            class="fishd-table-content"
          >
            <div
              class="fishd-table-scroll"
            >
              <div
                class="fishd-table-body"
              >
                <table
                  class=""
                >
                  <colgroup>
                    <col />
                    <col />
                  </colgroup>
                  <thead
                    class="fishd-table-thead"
                  >
                    <tr>
                      <th
                        class="fishd-table-fixed-columns-in-body fishd-table-selection-column"
                      >
                        <span>
                          <div
                            class="fishd-table-selection"
                          >
                            <label
                              class="fishd-checkbox-wrapper"
                            >
                              <span
                                class="fishd-checkbox"
                              >
                                <input
                                  class="fishd-checkbox-input"
                                  type="checkbox"
                                />
                                <span
                                  class="fishd-checkbox-inner"
                                />
                              </span>
                            </label>
                          </div>
                        </span>
                      </th>
                      <th
                        class=""
                      >
                        <span>
                          Name
                        </span>
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="fishd-table-tbody"
                  >
                    <tr
                      class="fishd-table-row fishd-table-row-level-0"
                      data-row-key="0"
                    >
                      <td
                        class="fishd-table-fixed-columns-in-body fishd-table-selection-column"
                      >
                        <span>
                          <label
                            class="fishd-checkbox-wrapper"
                          >
                            <span
                              class="fishd-checkbox"
                            >
                              <input
                                class="fishd-checkbox-input"
                                type="checkbox"
                              />
                              <span
                                class="fishd-checkbox-inner"
                              />
                            </span>
                          </label>
                        </span>
                      </td>
                      <td
                        class=""
                      >
                        <span
                          class="fishd-table-row-indent indent-level-0"
                          style="padding-left:0px"
                        />
                        Jack
                      </td>
                    </tr>
                    <tr
                      class="fishd-table-row fishd-table-row-level-0"
                      data-row-key="1"
                    >
                      <td
                        class="fishd-table-fixed-columns-in-body fishd-table-selection-column"
                      >
                        <span>
                          <label
                            class="fishd-checkbox-wrapper"
                          >
                            <span
                              class="fishd-checkbox"
                            >
                              <input
                                class="fishd-checkbox-input"
                                type="checkbox"
                              />
                              <span
                                class="fishd-checkbox-inner"
                              />
                            </span>
                          </label>
                        </span>
                      </td>
                      <td
                        class=""
                      >
                        <span
                          class="fishd-table-row-indent indent-level-0"
                          style="padding-left:0px"
                        />
                        Lucy
                      </td>
                    </tr>
                    <tr
                      class="fishd-table-row fishd-table-row-level-0"
                      data-row-key="2"
                    >
                      <td
                        class="fishd-table-fixed-columns-in-body fishd-table-selection-column"
                      >
                        <span>
                          <label
                            class="fishd-checkbox-wrapper"
                          >
                            <span
                              class="fishd-checkbox"
                            >
                              <input
                                class="fishd-checkbox-input"
                                type="checkbox"
                              />
                              <span
                                class="fishd-checkbox-inner"
                              />
                            </span>
                          </label>
                        </span>
                      </td>
                      <td
                        class=""
                      >
                        <span
                          class="fishd-table-row-indent indent-level-0"
                          style="padding-left:0px"
                        />
                        Tom
                      </td>
                    </tr>
                    <tr
                      class="fishd-table-row fishd-table-row-level-0"
                      data-row-key="3"
                    >
                      <td
                        class="fishd-table-fixed-columns-in-body fishd-table-selection-column"
                      >
                        <span>
                          <label
                            class="fishd-checkbox-wrapper"
                          >
                            <span
                              class="fishd-checkbox"
                            >
                              <input
                                class="fishd-checkbox-input"
                                type="checkbox"
                              />
                              <span
                                class="fishd-checkbox-inner"
                              />
                            </span>
                          </label>
                        </span>
                      </td>
                      <td
                        class=""
                      >
                        <span
                          class="fishd-table-row-indent indent-level-0"
                          style="padding-left:0px"
                        />
                        Jerry
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <div
              class="fishd-table-fixed-left"
            >
              <div
                class="fishd-table-body-outer"
                style="-webkit-transform:translate3d (0, 0, 0)"
              >
                <div
                  class="fishd-table-body-inner"
                >
                  <table
                    class="fishd-table-fixed"
                  >
                    <colgroup>
                      <col />
                    </colgroup>
                    <thead
                      class="fishd-table-thead"
                    >
                      <tr>
                        <th
                          class="fishd-table-selection-column"
                        >
                          <span>
                            <div
                              class="fishd-table-selection"
                            >
                              <label
                                class="fishd-checkbox-wrapper"
                              >
                                <span
                                  class="fishd-checkbox"
                                >
                                  <input
                                    class="fishd-checkbox-input"
                                    type="checkbox"
                                  />
                                  <span
                                    class="fishd-checkbox-inner"
                                  />
                                </span>
                              </label>
                            </div>
                          </span>
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="fishd-table-tbody"
                    >
                      <tr
                        class="fishd-table-row fishd-table-row-level-0"
                        data-row-key="0"
                      >
                        <td
                          class="fishd-table-selection-column"
                        >
                          <span>
                            <label
                              class="fishd-checkbox-wrapper"
                            >
                              <span
                                class="fishd-checkbox"
                              >
                                <input
                                  class="fishd-checkbox-input"
                                  type="checkbox"
                                />
                                <span
                                  class="fishd-checkbox-inner"
                                />
                              </span>
                            </label>
                          </span>
                        </td>
                      </tr>
                      <tr
                        class="fishd-table-row fishd-table-row-level-0"
                        data-row-key="1"
                      >
                        <td
                          class="fishd-table-selection-column"
                        >
                          <span>
                            <label
                              class="fishd-checkbox-wrapper"
                            >
                              <span
                                class="fishd-checkbox"
                              >
                                <input
                                  class="fishd-checkbox-input"
                                  type="checkbox"
                                />
                                <span
                                  class="fishd-checkbox-inner"
                                />
                              </span>
                            </label>
                          </span>
                        </td>
                      </tr>
                      <tr
                        class="fishd-table-row fishd-table-row-level-0"
                        data-row-key="2"
                      >
                        <td
                          class="fishd-table-selection-column"
                        >
                          <span>
                            <label
                              class="fishd-checkbox-wrapper"
                            >
                              <span
                                class="fishd-checkbox"
                              >
                                <input
                                  class="fishd-checkbox-input"
                                  type="checkbox"
                                />
                                <span
                                  class="fishd-checkbox-inner"
                                />
                              </span>
                            </label>
                          </span>
                        </td>
                      </tr>
                      <tr
                        class="fishd-table-row fishd-table-row-level-0"
                        data-row-key="3"
                      >
                        <td
                          class="fishd-table-selection-column"
                        >
                          <span>
                            <label
                              class="fishd-checkbox-wrapper"
                            >
                              <span
                                class="fishd-checkbox"
                              >
                                <input
                                  class="fishd-checkbox-input"
                                  type="checkbox"
                                />
                                <span
                                  class="fishd-checkbox-inner"
                                />
                              </span>
                            </label>
                          </span>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul
        class="fishd-pagination fishd-table-pagination"
        unselectable="unselectable"
      >
        <li
          aria-disabled="true"
          class="fishd-pagination-disabled fishd-pagination-prev"
          title="上一页"
        >
          <a
            class="fishd-pagination-item-link"
          />
        </li>
        <li
          class="fishd-pagination-item fishd-pagination-item-1 fishd-pagination-item-active"
          tabindex="0"
          title="1"
        >
          <a>
            1
          </a>
        </li>
        <li
          aria-disabled="true"
          class="fishd-pagination-disabled fishd-pagination-next"
          title="下一页"
        >
          <a
            class="fishd-pagination-item-link"
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Table.rowSelection render with default selection correctly 1`] = `
<div
  class="fishd-table-wrapper"
>
  <div
    class="fishd-spin-nested-loading"
  >
    <div
      class="fishd-spin-nested"
    >
      <div>
        <div
          class="fishd-table fishd-table-default fishd-table-scroll-position-left"
        >
          <div
            class="fishd-table-content"
          >
            <div
              class="fishd-table-body"
            >
              <table
                class=""
              >
                <colgroup>
                  <col />
                  <col />
                </colgroup>
                <thead
                  class="fishd-table-thead"
                >
                  <tr>
                    <th
                      class="fishd-table-selection-column fishd-table-selection-column-custom"
                    >
                      <span>
                        <div
                          class="fishd-table-selection"
                        >
                          <label
                            class="fishd-table-selection-select-all-custom fishd-checkbox-wrapper"
                          >
                            <span
                              class="fishd-checkbox"
                            >
                              <input
                                class="fishd-checkbox-input"
                                type="checkbox"
                              />
                              <span
                                class="fishd-checkbox-inner"
                              />
                            </span>
                          </label>
                          <div
                            class="fishd-table-selection-down fishd-dropdown-trigger"
                          >
                            <i
                              class="fishdicon fishdicon-down-bolder"
                            />
                          </div>
                        </div>
                      </span>
                    </th>
                    <th
                      class=""
                    >
                      <span>
                        Name
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="fishd-table-tbody"
                >
                  <tr
                    class="fishd-table-row fishd-table-row-level-0"
                    data-row-key="0"
                  >
                    <td
                      class="fishd-table-selection-column fishd-table-selection-column-custom"
                    >
                      <span>
                        <label
                          class="fishd-checkbox-wrapper"
                        >
                          <span
                            class="fishd-checkbox"
                          >
                            <input
                              class="fishd-checkbox-input"
                              type="checkbox"
                            />
                            <span
                              class="fishd-checkbox-inner"
                            />
                          </span>
                        </label>
                      </span>
                    </td>
                    <td
                      class=""
                    >
                      <span
                        class="fishd-table-row-indent indent-level-0"
                        style="padding-left:0px"
                      />
                      Jack
                    </td>
                  </tr>
                  <tr
                    class="fishd-table-row fishd-table-row-level-0"
                    data-row-key="1"
                  >
                    <td
                      class="fishd-table-selection-column fishd-table-selection-column-custom"
                    >
                      <span>
                        <label
                          class="fishd-checkbox-wrapper"
                        >
                          <span
                            class="fishd-checkbox"
                          >
                            <input
                              class="fishd-checkbox-input"
                              type="checkbox"
                            />
                            <span
                              class="fishd-checkbox-inner"
                            />
                          </span>
                        </label>
                      </span>
                    </td>
                    <td
                      class=""
                    >
                      <span
                        class="fishd-table-row-indent indent-level-0"
                        style="padding-left:0px"
                      />
                      Lucy
                    </td>
                  </tr>
                  <tr
                    class="fishd-table-row fishd-table-row-level-0"
                    data-row-key="2"
                  >
                    <td
                      class="fishd-table-selection-column fishd-table-selection-column-custom"
                    >
                      <span>
                        <label
                          class="fishd-checkbox-wrapper"
                        >
                          <span
                            class="fishd-checkbox"
                          >
                            <input
                              class="fishd-checkbox-input"
                              type="checkbox"
                            />
                            <span
                              class="fishd-checkbox-inner"
                            />
                          </span>
                        </label>
                      </span>
                    </td>
                    <td
                      class=""
                    >
                      <span
                        class="fishd-table-row-indent indent-level-0"
                        style="padding-left:0px"
                      />
                      Tom
                    </td>
                  </tr>
                  <tr
                    class="fishd-table-row fishd-table-row-level-0"
                    data-row-key="3"
                  >
                    <td
                      class="fishd-table-selection-column fishd-table-selection-column-custom"
                    >
                      <span>
                        <label
                          class="fishd-checkbox-wrapper"
                        >
                          <span
                            class="fishd-checkbox"
                          >
                            <input
                              class="fishd-checkbox-input"
                              type="checkbox"
                            />
                            <span
                              class="fishd-checkbox-inner"
                            />
                          </span>
                        </label>
                      </span>
                    </td>
                    <td
                      class=""
                    >
                      <span
                        class="fishd-table-row-indent indent-level-0"
                        style="padding-left:0px"
                      />
                      Jerry
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
      <ul
        class="fishd-pagination fishd-table-pagination"
        unselectable="unselectable"
      >
        <li
          aria-disabled="true"
          class="fishd-pagination-disabled fishd-pagination-prev"
          title="上一页"
        >
          <a
            class="fishd-pagination-item-link"
          />
        </li>
        <li
          class="fishd-pagination-item fishd-pagination-item-1 fishd-pagination-item-active"
          tabindex="0"
          title="1"
        >
          <a>
            1
          </a>
        </li>
        <li
          aria-disabled="true"
          class="fishd-pagination-disabled fishd-pagination-next"
          title="下一页"
        >
          <a
            class="fishd-pagination-item-link"
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;
